### DISCUSSION TITLE
p-refinement

### DISCUSSION THREAD

**Discussion by whiffee** [idaholab/moose#18984] [Q&A Modules: General] [answered]
Hi All,

I see in Moose example ex05 mesh adaptivity is demonstrated, but I can't tell whether it is h- or p-refinement.
I notice that in the examples in the libMesh repository the distinction is clearly stated. Does/can Moose take 
advantage of this choice? How would that change the input file?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Moose does h-refinement with the unifom_refine options (or with adaptivity). Uniform refinement can be used by specifying `-r xx` on the command line or by adding `uniform_refine = xx` to the [Mesh] block. 
Adaptivity and mesh generators allow you to tune the mesh refinement more than just a uniform refinement. 
You can do p-refinement manually by changing the FE family order in the [Variables] block. 

Guillaume

**Comment by whiffee** [comment]
I'm sort of stuck with the mesh I have, a hex mesh. I've tried .inp import and it seems to work.

I realize I probably wasn't ready to ask a question about this. However, since I already have,
I'm supposing a variables block like

[Variables]
    
      order = FOURTH
      family = LAGRANGE

[]

would turn on p-refinement, with a maximum basis order of 4. But how would the decision be made
as to which element quality would be assigned which order?

thanks for your patience
Gary

**Reply by GiudGiud** [reply]
moose can refine that mesh (h-refinement) for you, with the options I listed. It doesnt matter that it's imported from a .inp file for refinement. 

This p-refinement would only work if the mesh can accomodate such a high order variable, eg if it has all the extra points in each element. 
@roystgnr will know more than me here

**Reply by whiffee** [reply]
Please don't think I'm knocking h-refinement. It's great if the local mesh quality can support it. I'm just fishing for information on p-refinement because Wikipedia said Moose could do it.

**Reply by roystgnr** [reply]
I know about the libMesh side of things, so we can start off there, at least:

To do adaptive p-refinement, you want your "base" variable type to be the lowest order you ever want to see.  If you ask for `FOURTH` order, that means that p=4 will be the *minimum* in your subsequent simulation, not the maximum.

We don't currently support p>2 with `LAGRANGE` (or `L2_LAGRANGE`, or `RATIONAL_BERNSTEIN`) elements, simply because we don't have any geometric Elem types with enough nodes to get cubics with an FE type that sort of inherently associates each Degree of Freedom with a specific node.  Adding higher order geometric elements (at least for the edge/quad/hex Elem types) is on my to-do list now but I'm not sure if I'll get to it by the end of the calendar year.

To get adaptive p refinement (not just uniform), you also currently need an FE type that's "hierarchic" in the way it assigns degrees of freedom.  Right now that limits you to `HERMITE`, `HIERARCHIC`, `L2_HIERARCHIC`, `MONOMIAL`, `MONOMIAL_VEC`, `SIDE_HIERARCHIC`, `SZABAB`, and/or `XYZ` elements.  Some of those elements have other quirks or limitations too - `L2_HIERARCHIC`, `MONOMIAL`s and `XYZ` are discontinuous, `SIDE_HIERARCHIC` is just defined on inter-element boundaries, etc.

So the variables block you'd want to start with would be something like
```
order = FIRST
family = HIERARCHIC
```

And then you'd do p adaptivity from there ... and I'm not actually sure how to do that in Moose.  In libMesh, after you've run an error estimator and used it to flag elements (for h refinement by default), you just run `MeshRefinement::switch_h_to_p_refinement()` to switch those "h refine me" flags to be "p refine me" flags before calling `refine_and_coarsen_elements`.  A `git grep` isn't showing me any calls to that switch from within Moose, though.  I'm not seeing anything obvious in Adaptivity.C either.  Do we not have any way to do this from a Moose input file, without making users write custom C++ modules?  If not then that ought to go on my to-do list ASAP; it'd be pretty easy to add a new Adaptivity option that would at least cover most use cases (probably anything without mesh displacement).

**Reply by YaqiWang** [reply]
Unrelated, about `L2_LAGRANGE`, was that improved to support p>2 without the geometric constraint?

**Reply by roystgnr** [reply]
No; I'm going to wait until I've got higher-order `LAGRANGE`.  I assume everyone is going to expect/insist on local DoF ordering matching between the two, and I'd hate to run into some catch with `LAGRANGE` and the higher-order node ordering and discover that I had to make a backwards-incompatible change to `L2_LAGRANGE` to match.

Also ... until we have higher-order geometric elements (at which point I'll be adding the higher-order `LAGRANGE` immediately; we need that for just default master->physical mappings), It's hard to see the point of higher-order `L2_LAGRANGE`.  I guess we'd effectively have "virtual nodes" that don't exist, to e.g. define 16-"node" `L2_LAGRANGE` shapes on a Quad9?  That's doable, but what would the benefit even be, over `L2_HIERARCHIC`?  Lagrange shape functions don't work great in the mixed-order setting - you can put a first-order FE on a second-order Elem easily enough since the nodes are subsets, but you can't do that in general, only in the `p1 == 2^n p2` case.  We may never support `LAGRANGE` on `log2(elem_p/p) \notin W` cases, because there's no Lagrange node on which to "put" some of the DoFs there.  Supporting `L2_LAGRANGE` for those cases would be easier, but it'd be hard to convince me it was worth the trouble.

**Reply by YaqiWang** [reply]
I think I used high order L2_LAGRANGE without changing the mesh (first order) with DFEM. Now I am not sure if things were working ok.

**Reply by roystgnr** [reply]
If you didn't use the Lagrange property in your code (i.e. if you could have just run with `L2_HIERARCHIC` instead), then that actually should have worked okay!  The trouble with high order `L2_LAGRANGE` is that one tends to naturally assume that `L2_LAGRANGE` elements are Lagrange elements, i.e. that `phi_i(node_j)` is 1 if `i==j` and 0 if `i!=j`.  With second-order `L2_LAGRANGE` on a first-order Elem, `node_j` doesn't even exist for high j.  But where it starts getting even worse is for third-order `L2_LAGRANGE` on a second-order Elem: there the first several non-vertex nodes do exist, but they're not at the 1/3 and 2/3 intervals where you'd expect them for a cubic, they're at the mid points suitable for a quadratic, and trying to evaluate `phi_i(node_j)` can actually give you wrong results for a cubic `phi_i` on a quadratic `node_j`, not just undefined-behavior-but-probably-a-nice-clean-segfault.

Anyway, if you want to chat more about this, let's open a new discussion entry, not clutter up comments on the p-refinement question.
